python/tank/bootstrap/threading.py

```diff
@@ -0,0 +1,273 @@
+        self._worker.progressing.connect(self._progress_bootstrap)
+
+        # Handle completion of the toolkit bootstrap by the worker.
+        self._worker.completed.connect(self._complete_bootstrap)
+
+        # Handle failure of the toolkit bootstrap by the worker.
+        self._worker.failed.connect(self._fail_bootstrap)
+
+    def set_callbacks(self, progress_callback, completed_callback, failed_callback):
+        """
+        Sets the callback functions used to signify the progress, the completion and the failure of the bootstrap.
+
+        These functions must have the following signatures:
+
+            ``progress_callback(step_number, message, current_index, maximum_index)``
+
+            where:
+            - ``step_number`` is the current progress step number,
+                              from 1 to ``ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+            - ``message`` is the progress message to report.
+            - ``current_index`` is an optional current item number being looped over.
+            - ``maximum_index`` is an optional maximum item number being looped over.
+`
+            ``completed_callback(engine)``
+
+            where:
+            - ``engine`` is the launched :class:`sgtk.platform.Engine` instance.
+
+            ``failed_callback(step, exception)``
+
+            where:
+            - ``step`` is the bootstrap step ("sgtk" or "engine") that raised the exception.
+            - ``exception`` is the python exception raised while bootstrapping.
+
+        :param progress_callback: Callback function that reports back on the toolkit and engine bootstrap progress.
+        :param completed_callback: Callback function that handles cleanup after successful completion of the bootstrap.
+        :param failed_callback: Callback function that handles cleanup after failed completion of the bootstrap.
+        """
+
+        self._progress_callback = progress_callback
+        self._completed_callback = completed_callback
+        self._failed_callback = failed_callback
+
+    def bootstrap(self):
+        """
+        Starts the asynchronous bootstrap logic.
+        """
+
+        # Start the QThread object event loop in its new thread context.
+        self._thread.start()
+
+    @QtCore.Slot(int, str, int, int)
+    def _progress_bootstrap(self, step_number, message, current_index, maximum_index):
+        """
+        Callback slot that reports back on the toolkit and engine bootstrap progress.
+
+        :param step_number: Current progress step number, from 1 to ``ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+        :param message: Progress message to report.
+        :param current_index: Optional current item number being looped over, or None.
+        :param maximum_index: Optional maximum item number being looped over, or None.
+        """
+
+        if self._progress_callback:
+            self._progress_callback(step_number, message, current_index, maximum_index)
+
+    @QtCore.Slot(Sgtk)
```

Comment:
As noted before, I used the decorators to shield against the slot threading issue described here:
http://stackoverflow.com/questions/20752154/pyqt-connecting-a-signal-to-a-slot-to-start-a-background-operation

    Also, it might be worth expanding further on the PyQt-specific issue. In PyQt, a signal can be connected to a Qt slot, another signal, or any python callable. For the latter case, a proxy object is created internally that wraps the python callable and provides the slot that is required by the Qt signal/slot mechanism.

    It is this proxy object that is the cause of the problem. Once the proxy is created, PyQt will simply do this:

    if (rx_qobj)
        proxy->moveToThread(rx_qobj->thread());

    which is fine if the connection is made after the receiving object has been moved to its thread; but if it's made before, the proxy will stay in the main thread.

    Using the @pyqtSlot decorator avoids this issue altogether, because it creates a Qt slot more directly and does not use a proxy object at all.

    Finally, it should also be noted that this issue does not currently affect PySide.